azurerm/internal/services/storage/blobs.go

```diff
@@ -0,0 +1,306 @@
+		return fmt.Errorf("Error splitting source file %q into pages: %s", sbu.Source, err)
+	}
+
+	// finally we upload the contents of said file
+	pages := make(chan storageBlobPage, len(pageList))
+	errors := make(chan error, len(pageList))
+	wg := &sync.WaitGroup{}
+	wg.Add(len(pageList))
+
+	total := int64(0)
+	for _, page := range pageList {
+		total += page.section.Size()
+		pages <- page
+	}
+	close(pages)
+
+	for i := 0; i < workerCount; i++ {
+		go sbu.blobPageUploadWorker(ctx, blobPageUploadContext{
+			blobSize: fileSize,
+			pages:    pages,
+			errors:   errors,
+			wg:       wg,
+		})
+	}
+
+	wg.Wait()
+
+	if len(errors) > 0 {
+		return fmt.Errorf("Error while uploading source file %q: %s", sbu.Source, <-errors)
+	}
+
+	return nil
+}
+
+const (
+	minPageSize int64 = 4 * 1024
+
+	// TODO: investigate whether this can be bumped to 100MB with the new API
+	maxPageSize int64 = 4 * 1024 * 1024
+)
+
+func (sbu BlobUpload) storageBlobPageSplit(file io.ReaderAt, fileSize int64) ([]storageBlobPage, error) {
+	// whilst the file Size can be any arbitrary Size, it must be uploaded in fixed-Size pages
+	blobSize := fileSize
+	if fileSize%minPageSize != 0 {
+		blobSize = fileSize + (minPageSize - (fileSize % minPageSize))
+	}
+
+	emptyPage := make([]byte, minPageSize)
+
+	type byteRange struct {
+		offset int64
+		length int64
+	}
+
+	var nonEmptyRanges []byteRange
+	var currentRange byteRange
+	for i := int64(0); i < blobSize; i += minPageSize {
+		pageBuf := make([]byte, minPageSize)
+		if _, err := file.ReadAt(pageBuf, i); err != nil && err != io.EOF {
+			return nil, fmt.Errorf("Could not read chunk at %d: %s", i, err)
+		}
+
+		if bytes.Equal(pageBuf, emptyPage) {
+			if currentRange.length != 0 {
+				nonEmptyRanges = append(nonEmptyRanges, currentRange)
+			}
+			currentRange = byteRange{
+				offset: i + minPageSize,
+			}
+		} else {
+			currentRange.length += minPageSize
+			if currentRange.length == maxPageSize || (currentRange.offset+currentRange.length == blobSize) {
+				nonEmptyRanges = append(nonEmptyRanges, currentRange)
+				currentRange = byteRange{
+					offset: i + minPageSize,
+				}
+			}
+		}
+	}
+
+	var pages []storageBlobPage
+	for _, nonEmptyRange := range nonEmptyRanges {
+		pages = append(pages, storageBlobPage{
+			offset:  nonEmptyRange.offset,
+			section: io.NewSectionReader(file, nonEmptyRange.offset, nonEmptyRange.length),
+		})
+	}
+
+	return pages, nil
+}
+
+type blobPageUploadContext struct {
+	blobSize int64
+	pages    chan storageBlobPage
+	errors   chan error
+	wg       *sync.WaitGroup
+}
+
+func (sbu BlobUpload) blobPageUploadWorker(ctx context.Context, uploadCtx blobPageUploadContext) {
+	for page := range uploadCtx.pages {
+		start := page.offset
+		end := page.offset + page.section.Size() - 1
+		if end > uploadCtx.blobSize-1 {
+			end = uploadCtx.blobSize - 1
+		}
+		size := end - start + 1
+
+		chunk := make([]byte, size)
+		_, err := page.section.Read(chunk)
+		if err != nil && err != io.EOF {
```

Comment:
These lines could be merged:
```suggestion
		if _, err := page.section.Read(chunk) != nil && err != io.EOF {
```